### PR TITLE
Test across versions and platforms

### LINKED ISSUE #185: Clean up environment file
Packages we can remove:
- numpy ??
- seaborn
- pysal
- cartopy
- sklearn
- geocoder 
- mne
- tqdm 
- geopy
- pyqt

## Things that could move to dev-requirements.txt:

- codecov
- pytest stuff could go in dev requirements files??
 - pytest
- pytest-cov
NOTE: would want to ensure they install properly using pip...

### LINKED ISSUE #213: Testing earthpy across Python versions
We want to ensure that earthpy works across different versions of python. @mbjoseph  will look into how we  can best implement testing

### PR DESCRIPTION
**Pull request by mbjoseph**
This tests earthpy against python 3.5 through 3.7 using tox, and also uses AppVeyor to test earthpy on Windows. Last, this updates documentation for development using conda or virtualenv and testing of multiple python versions with tox.

As part of the review, let's take the updated development instructions in `CONTRIBUTING.rst` for a test drive. I will be curious to see what issues come up, especially on Windows. I'm opening this as a draft PR until we get some more eyes on it.

Solves the hardest part of #228 - testing on Windows (but we should leave that issue open until we get some tests in CI for macOS). Closes #213. Also closes #185.

### CODE REVIEW

**Comment by mbjoseph** [comment]
@lwasser and @joemcglinchy I would love if you could both review this PR.

As part of your reviews, please go through the updated instructions in `CONTRIBUTING.rst`, noting the changes to the creation of the `earthpy-dev` environment. I suspect this will not work on Windows, in which case we can modify the instructions to tell users to create the `earthpy-dev` environment using the `environment.yml` file (as we do currently).

**Comment by joemcglinchy** [comment]
everything builds OK on my Windows 10 machine!

**Comment by mbjoseph** [comment]
Thanks @joemcglinchy! I have revised the instructions to have Windows users create the environment from `environment.yml`. Can you retry following the updated instructions? 

TLDR:

```bash
conda env create -f environment.yml
source activate earthpy-dev
pip install -e .
pip install -r dev-requirements.txt
pre-commit install
pytest
make docs
```

I wouldn't bother trying to run `tox` on Windows, but if you have any additional instructions that might be useful for Windows users to install Make - that would be super useful. Feel free to push to this branch or respond on this thread if there's something you would add to what I have.

File: CONTRIBUTING.rst [hunk 1]
```diff
@@ -24,42 +22,117 @@ If you are proposing a feature:
...
+
+Using conda, there are two options.
+
+1. The first and easiest option is to create an environment from the
+``environment.yml`` file.
+Note that this will only allow you to test against one version of python
+locally, but this is the recommended option on Windows::
+
     $ conda env create -f environment.yml
     $ source activate earthpy-dev
```
Comment (lwasser) [inline CONTRIBUTING.rst]: Heya @mbjoseph so interestingly we learned yesterday that the new versions of conda all use `conda activate` rather than `source activate` ALSO this works on windows EVEN in command prompt after testing :) however i am not sure how far back it goes ie whether older versions of conda support it but apparently they are deprecating this.  [More here](https://conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html)

**Comment by lwasser** [comment]
it looks like that `libspatialindex_c` is an issue??

**Comment by lwasser** [comment]
ok i used 
`brew install spatialindex` this worked well @mbjoseph  but now i'm here
```
____________________________________ ERROR collecting earthpy/tests/test_plot.py ____________________________________
ImportError while importing test module '/Users/leah-su/Documents/github/1-python/earthpy/earthpy/tests/test_plot.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
earthpy/tests/test_plot.py:5: in <module>
    import matplotlib.pyplot as plt
../../../../anaconda3/envs/earthpy-dev/lib/python3.7/site-packages/matplotlib/pyplot.py:2374: in <module>
    switch_backend(rcParams["backend"])
../../../../anaconda3/envs/earthpy-dev/lib/python3.7/site-packages/matplotlib/pyplot.py:207: in switch_backend
    backend_mod = importlib.import_module(backend_name)
../../../../anaconda3/envs/earthpy-dev/lib/python3.7/importlib/__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
../../../../anaconda3/envs/earthpy-dev/lib/python3.7/site-packages/matplotlib/backends/backend_macosx.py:14: in <module>
    from matplotlib.backends import _macosx
E   ImportError: Python is not installed as a framework. The Mac OS X backend will not be able to function correctly if Python is not installed as a framework. See the Python documentation for more information on installing Python as a framework on Mac OS X. Please either reinstall Python as a framework, or try one of the other backends. If you are using (Ana)Conda please install python.app and replace the use of 'python' with 'pythonw'. See 'Working with Matplotlib on OSX' in the Matplotlib FAQ for more information.
______________________________ ERROR collecting earthpy/tests/test_plot_draw_legend.py ______________________________
ImportError while importing test module '/Users/leah-su/Documents/github/1-python/earthpy/earthpy/tests/test_plot_draw_legend.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
earthpy/tests/test_plot_draw_legend.py:6: in <module>
    import matplotlib.pyplot as plt
../../../../anaconda3/envs/earthpy-dev/lib/python3.7/site-packages/matplotlib/pyplot.py:2374: in <module>
    switch_backend(rcParams["backend"])
../../../../anaconda3/envs/earthpy-dev/lib/python3.7/site-packages/matplotlib/pyplot.py:207: in switch_backend
    backend_mod = importlib.import_module(backend_name)
../../../../anaconda3/envs/earthpy-dev/lib/python3.7/importlib/__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
../../../../anaconda3/envs/earthpy-dev/lib/python3.7/site-packages/matplotlib/backends/backend_macosx.py:14: in <module>
    from matplotlib.backends import _macosx
E   ImportError: Python is not installed as a framework. The Mac OS X backend will not be able to function correctly if Python is not installed as a framework. See the Python documentation for more information on installing Python as a framework on Mac OS X. Please either reinstall Python as a framework, or try one of the other backends. If you are using (Ana)Conda please install python.app and replace the use of 'python' with 'pythonw'. See 'Working with Matplotlib on OSX' in the Matplotlib FAQ for more information.
__________________________________ ERROR collecting earthpy/tests/test_plot_rgb.py __________________________________
ImportError while importing test module '/Users/leah-su/Documents/github/1-python/earthpy/earthpy/tests/test_plot_rgb.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
earthpy/tests/test_plot_rgb.py:5: in <module>
    import matplotlib.pyplot as plt
../../../../anaconda3/envs/earthpy-dev/lib/python3.7/site-packages/matplotlib/pyplot.py:2374: in <module>
    switch_backend(rcParams["backend"])
../../../../anaconda3/envs/earthpy-dev/lib/python3.7/site-packages/matplotlib/pyplot.py:207: in switch_backend
    backend_mod = importlib.import_module(backend_name)
../../../../anaconda3/envs/earthpy-dev/lib/python3.7/importlib/__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
../../../../anaconda3/envs/earthpy-dev/lib/python3.7/site-packages/matplotlib/backends/backend_macosx.py:14: in <module>
    from matplotlib.backends import _macosx
E   ImportError: Python is not installed as a framework. The Mac OS X backend will not be able to function correctly if Python is not installed as a framework. See the Python documentation for more information on installing Python as a framework on Mac OS X. Please either reinstall Python as a framework, or try one of the other backends. If you are using (Ana)Conda please install python.app and replace the use of 'python' with 'pythonw'. See 'Working with Matplotlib on OSX' in the Matplotlib FAQ for more information.
================================================= warnings summary ==================================================
/Users/leah-su/anaconda3/envs/earthpy-dev/lib/python3.7/site-packages/rasterio/crs.py:21
  /Users/leah-su/anaconda3/envs/earthpy-dev/lib/python3.7/site-packages/rasterio/crs.py:21: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated, and in 3.8 it will stop working
    class CRS(collections.Mapping):

-- Docs: https://docs.pytest.org/en/latest/warnings.html
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 3 errors during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
======================================== 1 warnings, 3 error in 3.32 seconds =====
```

**Comment by joemcglinchy** [comment]
@mbjoseph your updated instructions work for windows. I do not have make installed... there is no windows equivalent. One would have to either install Cygwin with `make` specified in the installation or GnuWin. What are the implications for note running `make docs`?

**Comment by mbjoseph** [comment]
Thanks @joemcglinchy for testing locally. For Make on Windows I was thinking about recommending something like this: http://gnuwin32.sourceforge.net/packages/make.htm or https://chocolatey.org/packages/make or https://github.com/swcarpentry/windows-installer

Are those viable options? I'm pretty naiive about what's possible on Windows.

The implications for not running `make docs` locally is that you would not be able to build the documentation, ensure that all examples run, or preview read the docs in a web browser.

**Comment by joemcglinchy** [comment]
got it. Personally, I use `chocolatey` and like it. `chocalatey install make` and then `make docs` builds successfully for me.

File: CONTRIBUTING.rst (hunk 1 above)
Comment (mbjoseph) [reply CONTRIBUTING.rst]: Thanks @lwasser - I have updated the contributing file to reflect the `conda activate` syntax

**Comment by lwasser** [comment]
ok an update. 
i used the windows instructions
```
conda env create -f environment.yml
source activate earthpy-dev
pip install -e .
pip install -r dev-requirements.txt
pre-commit install
pytest
make docs
```
and things worked ok except pytest. i got this:
```
collected 83 items                                                                                                      

earthpy/tests/test__stack_bands.py .                                                                              [  1%]
earthpy/tests/test_clip.py .........                                                                              [ 12%]
earthpy/tests/test_io.py ........                                                                                 [ 21%]
earthpy/tests/test_mask.py ........                                                                               [ 31%]
earthpy/tests/test_plot.py ..Abort trap: 6
```

this is pretty weird as i haven't seen this before. normally i see failures.

**Comment by mbjoseph** [comment]
Very interesting @lwasser - the third test in `test_plot.py` is the first test in that module that makes a plot. Maybe this is a matplotlib thing.

Based on https://github.com/matplotlib/matplotlib/issues/13096, it seems like matplotlib is close to having a fix in place for OS X. 

If I'm correct in attributing the `Abort trap` error, it should arise with our current instructions, independently of the changes made on this PR. @lwasser do you run into the same issue when following the instructions that we currently have already on the master branch to make a fresh `earthpy-dev` environment? If so, then we should probably treat this as a separate issue, unrelated to this PR.

If you could 

1. try this workaround: https://github.com/ContinuumIO/anaconda-issues/issues/10554, and
2. see whether you get the same error using our current instructions on the master branch (ignoring everything in this PR)

then maybe we can get a better idea of what might be happening.

**Comment by mbjoseph** [comment]
@lwasser any updates on whether you run into the same issue with our current instructions?

**Comment by lwasser** [comment]
@mbjoseph  i wasn't able to get the test across versions install to work. However i was able to get the normal earthpy-dev envt with the black hooks and pytest to work properly.

**Comment by mbjoseph** [comment]
Thanks - does that mean you are not getting the `Abort trap` error when using the earthpy-dev environment? 

If the `earthpy-dev` environment is working for you, then we have it working on Linux (me locally), Windows (@joemcglinchy), and OS X (@lwasser).

**Comment by lwasser** [comment]
let me try again just for good measure.

**Comment by lwasser** [comment]
yup @mbjoseph the vanilla install of earthpy-dev and all of the requirements works just fine on my mac. the across versions install seems tricky but the basic dev install works just fine.

**Comment by mbjoseph** [comment]
Great, then I think this is ready for a review @lwasser!

**Comment by mbjoseph** [comment]
Ok, I merged the changes from master, and this time this one is actually ready for review @lwasser! 

As a side note, we had some code that merged yesterday that was incompatible with python 3.5 (fixed by https://github.com/earthlab/earthpy/pull/244/commits/95f43a2379fdca1fa0b9fa4ce5bbca079be25e32), so this PR is already paying off!

**Comment by lwasser** [comment]
@mbjoseph i've noticed you're still adding to this PR. do you want me to review it tomorrow? I've been waiting for the commits to stop :)

**Comment by mbjoseph** [comment]
It's ready for review @lwasser - the last commit was to fix a conflicting file in the changelog.

**Comment by lwasser** [comment]
ok great. i'll have a look tomorrow! thanks @mbjoseph

File: CONTRIBUTING.rst [hunk 2]
```diff
@@ -25,42 +23,118 @@ If you are proposing a feature:
...
+and click the **Fork** button in the top-right corner of the page.
+
+2. Clone your fork locally
+--------------------------
 
-3. Set up your fork for local development with conda::
+Use ``git clone`` to get a local copy of your EarthPy repository on your
+local filesystem::
 
+    $ git clone git@github.com:your_name_here/earthpy.git
```
Comment (lwasser) [inline CONTRIBUTING.rst]: hey @mbjoseph do most people use ssh? i don't ! maybe i should

**Comment by lwasser** [comment]
ok @mbjoseph this looks good! i have one question about cloning using ssh vs the regular way. I guess it's not a big deal but some people won't understand that is using SSH. 

ALso just a note that i have to always run `make docs -B` to get docs to build. Should we go ahead and get this puppy merged??

File: CONTRIBUTING.rst (hunk 2 above)
Comment (mbjoseph) [reply CONTRIBUTING.rst]: Not sure - we can always add instructions to clone both ways later on!

**Comment by mbjoseph** [comment]
Sounds good @lwasser! Thanks!
